File: projects/chess/chess2/js/lichess.js
// Lichess's game-analysis maths, ported line for line from lila so the numbers
// match lichess.org: accuracy (AccuracyPercent.scala) and move judgements
// (Advice.scala). Pure functions, no DOM — tests/lichess.test.js checks both
// against lichess.org's own analysis of real games.
//
// An eval is { cp } or { mate } from White's point of view; mate is never 0
// (a checkmated position has no eval, exactly as in lila).

const CEILING = 1000;             // Cp.CEILING
export const INITIAL = { cp: 15 };  // Cp.initial: the start position's eval

// Mates count as the ceiling (lila's forceAsCp followed by the cap)
const cpOf = e => e.mate !== undefined ? (e.mate > 0 ? CEILING : -CEILING) : e.cp;
const clamp = (x, lo, hi) => Math.max(lo, Math.min(hi, x));

// [-1, 1]; lila deliberately leaves cp uncapped here (Advice uses it raw)
const winningChances = cp => clamp(2 / (1 + Math.exp(-0.00368208 * cp)) - 1, -1, 1);

// [0, 100], White's point of view (WinPercent.fromCentiPawns caps first)
export const winPercent = e => 50 + 50 * winningChances(clamp(cpOf(e), -CEILING, CEILING));

function moveAccuracy(before, after) {
  if (after >= before) return 100;
  const raw = 103.1668100711649 * Math.exp(-0.04354415386753951 * (before - after)) - 3.166924740191411;
  return clamp(raw + 1, 0, 100);  // +1: uncertainty bonus for imperfect analysis
}

const mean = xs => xs.reduce((a, b) => a + b, 0) / xs.length;
const stdev = xs => { const m = mean(xs); return Math.sqrt(mean(xs.map(x => (x - m) ** 2))); };

// Game accuracy per colour from the evals after each move (not the start
// position). Mean of a volatility-weighted mean and a harmonic mean of the
// per-move accuracies. Returns { white, black } unrounded; null if no moves.
export function accuracy(evals, whiteStarts = true) {
  const wins = [INITIAL, ...evals].map(winPercent);
  const size = clamp(Math.floor(evals.length / 10), 2, 8);
  const windows = [];
  for (let i = 0; i < Math.min(size, wins.length) - 2; i++) windows.push(wins.slice(0, size));
  for (let i = 0; i + size <= wins.length; i++) windows.push(wins.slice(i, i + size));
  if (wins.length < size) windows.push(wins);  // Scala's sliding on a short list
  const weights = windows.map(w => clamp(stdev(w), 0.5, 12));

  const by = { white: [], black: [] };
  for (let i = 0; i < wins.length - 1 && i < weights.length; i++) {
    const white = (i % 2 === 0) === whiteStarts;
    const [prev, next] = [wins[i], wins[i + 1]];
    by[white ? 'white' : 'black'].push([white ? moveAccuracy(prev, next) : moveAccuracy(next, prev), weights[i]]);
  }
  const colour = moves => {
    if (!moves.length) return null;
    const weighted = moves.reduce((s, [a, w]) => s + a * w, 0) / moves.reduce((s, [, w]) => s + w, 0);
    const harmonic = moves.length / moves.reduce((s, [a]) => s + 1 / Math.max(1, a), 0);
    return (weighted + harmonic) / 2;
  };
  return evals.length ? { white: colour(by.white), black: colour(by.black) } : null;
}

// Judgement of one move from the evals before and after it, or null. lila only
// asks when the move differs from the engine's best — the caller checks that.
export function judge(prev, next, whiteMoved) {
  const pov = whiteMoved ? 1 : -1;
  if (prev.cp !== undefined && next.cp !== undefined) {
    const drop = (winningChances(prev.cp) - winningChances(next.cp)) * pov;
    return drop >= 0.3 ? 'blunder' : drop >= 0.2 ? 'mistake' : drop >= 0.1 ? 'inaccuracy' : null;
  }
  // Mate sequences, from the mover's point of view
  const pm = prev.mate !== undefined ? prev.mate * pov : undefined;
  const nm = next.mate !== undefined ? next.mate * pov : undefined;
  if (pm === undefined && nm < 0) {         // walked into a forced mate
    const cp = prev.cp * pov;
    return cp < -999 ? 'inaccuracy' : cp < -700 ? 'mistake' : 'blunder';
  }
  if (pm > 0 && (nm === undefined || nm < 0)) {  // let a forced mate slip
    const cp = nm === undefined ? next.cp * pov : 0;
    return cp > 999 ? 'inaccuracy' : cp > 700 ? 'mistake' : 'blunder';
  }
  return null;
}

File: projects/chess/coach/lichess.py
"""Lichess's analysis maths, the Python twin of chess2's js/lichess.js: accuracy
(lila AccuracyPercent) and move judgements (lila Advice). Both are checked
against lichess.org's own analysis of the same 12 games (tests/test_lichess.py).

An eval is {"cp": n} or {"mate": n}, White's point of view; mate is never 0.
"""
from __future__ import annotations

import math

CEILING = 1000
INITIAL = {"cp": 15}


def _cp(e: dict) -> int:
    return (CEILING if e["mate"] > 0 else -CEILING) if "mate" in e else e["cp"]


def _chances(cp: float) -> float:  # [-1, 1]; uncapped here, as Advice uses it
    return max(-1.0, min(1.0, 2 / (1 + math.exp(-0.00368208 * cp)) - 1))


def win_percent(e: dict) -> float:  # [0, 100], White's point of view
    return 50 + 50 * _chances(max(-CEILING, min(CEILING, _cp(e))))


def _move_accuracy(before: float, after: float) -> float:
    if after >= before:
        return 100.0
    raw = 103.1668100711649 * math.exp(-0.04354415386753951 * (before - after)) - 3.166924740191411
    return max(0.0, min(100.0, raw + 1))


def _stdev(xs: list[float]) -> float:
    m = sum(xs) / len(xs)
    return math.sqrt(sum((x - m) ** 2 for x in xs) / len(xs))


def accuracy(evals: list[dict], white_starts: bool = True) -> dict | None:
    """{white, black} accuracy from the evals after each move (unrounded)."""
    if not evals:
        return None
    wins = [win_percent(e) for e in [INITIAL, *evals]]
    size = max(2, min(8, len(evals) // 10))
    windows = [wins[:size]] * (min(size, len(wins)) - 2) + [wins[i:i + size] for i in range(len(wins) - size + 1)]
    weights = [max(0.5, min(12.0, _stdev(w))) for w in windows]
    by = {"white": [], "black": []}
    for i in range(min(len(wins) - 1, len(weights))):
        white = (i % 2 == 0) == white_starts
        prev, nxt = wins[i], wins[i + 1]
        by["white" if white else "black"].append(
            (_move_accuracy(prev, nxt) if white else _move_accuracy(nxt, prev), weights[i]))

    def colour(moves):
        if not moves:
            return None
        weighted = sum(a * w for a, w in moves) / sum(w for _, w in moves)
        harmonic = len(moves) / sum(1 / max(1.0, a) for a, _ in moves)
        return (weighted + harmonic) / 2
    return {c: colour(m) for c, m in by.items()}


def judge(prev: dict, nxt: dict, white_moved: bool) -> str | None:
    """'inaccuracy' | 'mistake' | 'blunder' | None. lila only asks when the move
    wasn't the engine's best; the caller checks that."""
    pov = 1 if white_moved else -1
    if "cp" in prev and "cp" in nxt:
        drop = (_chances(prev["cp"]) - _chances(nxt["cp"])) * pov
        return "blunder" if drop >= 0.3 else "mistake" if drop >= 0.2 else "inaccuracy" if drop >= 0.1 else None
    pm = prev["mate"] * pov if "mate" in prev else None
    nm = nxt["mate"] * pov if "mate" in nxt else None
    if pm is None and nm is not None and nm < 0:  # walked into a forced mate
        cp = prev["cp"] * pov
        return "inaccuracy" if cp < -999 else "mistake" if cp < -700 else "blunder"
    if pm is not None and pm > 0 and (nm is None or nm < 0):  # let a forced mate slip
        cp = nxt["cp"] * pov if nm is None else 0
        return "inaccuracy" if cp > 999 else "mistake" if cp > 700 else "blunder"
    return None
